### ISSUE TITLE
Poor performance on Android

### ISSUE THREAD

**Issue by buunguyen** [jbavari/cordova-plugin-video-editor#69]
I'm seeing really bad performance on a fast device (Nexus 6P). Transcoding a 3MB 3-second video file (with MPEG4 output, OptimizedForNetworkUse.YES and default settings for all others) takes almost a minute.

Is there any way to speed up the process? For example, if I explicitly specify width, height, bitrate etc. what should I use to trade file size for performance (i.e. large file is okay if performance is faster)?

**Comment by rossmartin** [comment]
I haven't tried too many options on android with transcoding I wish I had a good answer for you.  I'm not blaming ffmpeg in general for why it is slow on android - I think the root of the problem is that ffmpeg is not optimized for the CPUs in mobile devices (ARM, etc).  I'd like to look into other ways to transcode video on android other than ffmpeg I just haven't had the time.  Figuring out how to speed up transcoding on android is certainly up there on list of improvements for his plugin.  Thanks for the input your numbers above are similar to what I see on my ancient Galaxy s2 which isn't right.

The plugin is so much better on iOS I hope you've had the chance to try it.  I've worked on this plugin mainly for a cross platform app of my own that I have primarily focused on iOS with.

**Comment by buunguyen** [comment]
Thanks for your prompt response. I use your plugin to complement the [media-capture plugin](https://github.com/apache/cordova-plugin-media-capture), which compresses captured videos on iOS but not Android. I hope you have a chance to improve the Android story once you have time.

**Comment by buunguyen** [comment]
Not sure if it helps, but I found a library that does hardware accelerated transcoding for Android. It's not a cordova plugin though. https://github.com/ypresto/android-transcoder

**Comment by rossmartin** [comment]
Thanks that looks promising I wish it had existed when I created the android version of the plugin.  I'll be taking a look at it soon.

If you're interested in a media picker for iOS that doesn't compress your selected video to 720p check out this one I made recently - [cordova-plugin-instagram-assets-picker](https://github.com/rossmartin/cordova-plugin-instagram-assets-picker).

I plan on making a cordova plugin for this really good image picker that uses the new Photo framework - [GMImagePicker](https://github.com/guillermomuntaner/GMImagePicker).  This will be a huge plus for cordova on iOS.

**Comment by buunguyen** [comment]
@rossmartin the assets-picker looks awesome! I wish I knew about it before, would have saved much time. I'll give it a try in the next opportunity.

Hope to hear the good news about speeding up Android transcoding. I'd prefer to use your plugin as a one-stop shop for cross-platform transcoding/trimming instead of dealing with multiple libraries.

**Comment by MT--** [comment]
I'm really glad this plugin works and works pretty well. The Android transcoding is extremely slow on this Nexus 5X that I am developing on. The CPU optimization comment seems to be true. What would be a good way to implement the library that @buunguyen suggested for Android?

**Comment by rossmartin** [comment]
@MT--  I've only briefly looked at the [https://github.com/ypresto/android-transcoder](android-transcoder) but I do plan on trying it out when I get time.  I'm not sure that you'll be able to get very granular with it as far as transcoding settings (width, height, bitrate, etc.)  It appeared that is uses preset profiles for transcoding, almost like Apple's `AVAssetExportSession`.  This may be a better route on android though.

I plan to get rid of ffprobe or make it optional on android because it is just too large of a binary for such little feature.  Plus I added a `getVideoInfo` for both platforms that is good enough.

**Comment by MT--** [comment]
@rossmartin that sounds fantastic, I started working on a plugin last night that uses the `android-transcoder`, but you will probably be able to crank it out faster than me. I was just baking in the video size I wanted for my app. I think you'll be able to provide a good range of presets in your version. Maybe leave `ffmpeg` as an option for those who want to get really specific with their options? `getVideoInfo` works really well, too! Thanks for the great plugin! 

P.S. I'll post a repo here if I get to a point where I feel comfortable sharing my plugin. (This is my first go at writing a Cordova plugin since there have always existed good solutions.)

**Comment by rossmartin** [comment]
@buunguyen @MT-- 

I want to give an update on this - I pulled down the https://github.com/ypresto/android-transcoder and ran the example on my android (Galaxy S2 on lollipop).  Unfortunately I am unable to get the library to transcode any recorded video (H.264, MPEG4) at any resolution.  I think it is because my android device is so old and the hardware isn't recording the video at the correct "baseline profile".  Basically all I see is `"Fatal error while transcoding, this might be invalid format or bug in engine or Android."` when trying to transcode a video.

Can anyone with a more modern android run the example app in the `android-transcoder` repo and confirm that transcoding works for them?  I'm blocked on improving the android version of this plugin right now because my android device is so old.

**Comment by roysG** [comment]
any updates about this bug?!

**Comment by rossmartin** [comment]
@roysG I would like to revise the `transcodeVideo` function on android to use the android-transcoder library.  Unfortunately my android device is too old and won't work with the android-transcoder library.  It's not a matter of software, my old S2 is on lollipop.  I've also tried an emulator with various settings but it simply won't work.

What might work is if someone can record a short video on a more modern android and send it to me.  Then I should be able to get the android-transcoder library to transcode.  Let me know if you're interested in trying this.

**Comment by rossmartin** [comment]
@roysG I forgot to mention - this issue is not a bug but will be resolved in time.  To anyone willing pull requests are welcome.

**Comment by roysG** [comment]
Of course, i am ready to assist, just tell me what you need, and i will do :)

**Comment by roysG** [comment]
Do you want that i will record short video that you will see the quality?

**Comment by roysG** [comment]
??

**Comment by rossmartin** [comment]
@roysG Sorry for the late response.  Yeah if you can send me a short video that is recorded by a modern android device.  My email is rmartin311 at gmail dot com.  You can send me an email with the video attached as long as it is under 20 MB otherwise provide a link to the video.  Thanks.

**Comment by roysG** [comment]
Of course!, the size is not bigger than 150kb with your compression plugin

**Comment by shahin8r** [comment]
Dying to know how the progress is on this. My app loves your work @rossmartin

**Comment by rossmartin** [comment]
Awesome @shahin8r. I received a video recorded by @roysG android device but I've yet to try it with the android-transcoder.  Been busy with my day job and everyday stuff.  Hopefully I get some time within the next few days to integrate android-transcoder.

**Comment by AymanFekri** [comment]
Thank you @rossmartin . , 
iam agree that ffmpeg is slow in android , but i think you can change
https://github.com/jbavari/cordova-plugin-video-editor/blob/master/src/android/android-ffmpeg-java/src/org/ffmpeg/android/FfmpegController.java

 
`
		if (in.videoCodec != null)
		{
			cmd.add(Argument.VIDEOCODEC);
			cmd.add(in.videoCodec);

			if (in.videoCodec == "libx264")
			{
				cmd.add("-preset");
				cmd.add("ultrafast"); // needed b/c libx264 doesn't utilize all CPU cores
			}
		}
`

To

`

		if (out.videoCodec != null)
		{
			cmd.add(Argument.VIDEOCODEC);
			cmd.add(in.videoCodec);

			if (out.videoCodec == "libx264")
			{
				cmd.add("-preset");
				cmd.add("ultrafast"); // needed b/c libx264 doesn't utilize all CPU cores
			}
		}
`

it's  going to have better performance .

**Comment by shahin8r** [comment]
@rossmartin any progress with `android-transcoder` yet?

**Comment by rossmartin** [comment]
@shahin8r Sorry I've been really busy and haven't got to it yet.

**Comment by siada** [comment]
@ergovia-mobile has forked this project and created a branch which implements the android-transcoder project. I've done some basic tests and it works perfectly, the speed improvements are massive!

**Comment by rossmartin** [comment]
@siada Thanks I'll take a look and get it integrated.  I won't be able to test it with my android device.

**Comment by shahin8r** [comment]
@rossmartin just let us know when your done integrating it and I'll for sure test it on my Android device.

**Comment by shahin8r** [comment]
@rossmartin Tried it now and the speed improvement are crazy! However it's not possible to do multiple compressions simultaneously. In iOS this works fine.

**Comment by rossmartin** [comment]
@shahin8r Its integrated in the master branch but I'm going through the changes and will be updating documentation.  There are options that have been removed/modified for `transcodeVideo` and `createThumbnail`.  I'd like to make the options as consistent as possible for both platforms.  Also the `trim` function is gone for android after removing ffmpeg.

There were some modifications made to the `android-transcoder` to support limiting a video to a specified `duration`.  The option for this was `videoDuration` but I've changed it to `duration` on `transcodeVideo`.  I've yet to go through and look at the `createThumbnail` in more detail but it has changed also.

**Comment by shahin8r** [comment]
@rossmartin I've noticed problems with setting width/height. I see the arguments are missing for the cordova plugin on android now but the `maxWidth` should be `maxHeight` since it's the height that is affected on the output video. And passing `maxWidth` with a value when calling the cordova plugin with `VideoEditor.transcodeVideo` crashes the app.

**Comment by rossmartin** [comment]
The `maxWidth` option will end up scaling the width and height of the output video.  In the case of a portrait video it will be `maxHeight`.  Either way it does appear the transcoding is hardcoded to scale a video to 16:9 aspect ratio by comparing the width/height wanted to 960x540.

![image](https://cloud.githubusercontent.com/assets/2498502/16080557/ff7e38ba-32ce-11e6-802a-a844905a2d9d.png)

In the screenshot above the `inputFormat` width is always 960 and the height is always 540.  It is then compared to what is now `maxWidth`.

I did revise how the scaling is done in the `createThumbnail` function.  In my perspective having a `shorterLength` option is not as clear as being able to pass a width/height and have it scale based off of that.

Thoughts?

**Comment by cmboros** [comment]
@rossmartin the speed is great however the success or error function is not trigger. 
The thumbnail is working but also here first recording it give this error createThumbnailError, err: java.lang.IllegalArgumentException after is working every time. 

Also the same code is used on iOS and is working without any issue.

I have test on the Android 4.4.2 and the model is BLU ADVANCE 4.0L, LG G5 and Samsung Galaxy s7 all device behave the same. I try with croswalk and without as well.

I have create a gist with my file https://gist.github.com/cmboros/e8756d605659538390cbda1b40455e69

Thanks

**Comment by cmboros** [comment]
Anyone know how to get the output to work? Or is not working on android? Thanks

**Comment by rossmartin** [comment]
@shahin8r Can you do some testing with the latest version?  Unfortunately there is only so much I can test with my old android device.

@cmboros You are using meteor right?  I know there are some quirks about file access on meteor.  Can you check out this second bullet point here - http://guide.meteor.com/1.3-migration.html#mobile

**Comment by cmboros** [comment]
@rossmartin the file access is not the problem. The thumbnail is created but VideoEditor.transcodeVideo is not returning anything. The success and the error are ignored. VideoEditor.createThumbnail is working and is returning the thumbnail from the video.

**Comment by rossmartin** [comment]
@cmboros I can't replicate the issue with the callbacks not firing.  Are you using the latest version (1.1.0)?

**Comment by cmboros** [comment]
@rossmartin I will post the new discussions on the issue I open. I will upgrade to 1.1.0 now is using 1.0.9 
Thanks

**Comment by dheerajchugh1** [comment]
Hi

First of all thanks and kudos to all the contributors who wrote this plug-in. 
Is this performance issue on android device has been solved ? I am using the version 1.1.1 but still observing that compression of video is extremely slow.  

Please help me on this 

Thanks

**Comment by Xwalk-Company** [comment]
+1

**Comment by Xwalk-Company** [comment]
if you need help we want to work with you

**Comment by siada** [comment]
I feel like this issue can be closed as the original is now resolved with the removal of ffmpeg. I've been using the changes implemented for several months now across a few thousand devices and am yet to have any problems.

**Comment by dheerajchugh303** [comment]
@siada : I am using the version 1.1.1 and still facing the terrible low speed in my android device(Mi note 3). Please correct me if i am doing anything wrong.

What are the steps i need to follow to fix this issue.

**Comment by Xwalk-Company** [comment]
Yes siada
Actually the latest version of the plugin no longer uses FFMPEG but it applies[ MediaTranscoder ](https://github.com/ypresto/android-transcoder)
No poor performance for me

**Comment by dheerajchugh303** [comment]
Hi 

I am using the new version of the plugin 1.1.2 and dis the benchmark

Video Size = 320 MB
Video size after comression = 84 MB
Video length (in time) = 1 min
Time taken for comression  = 7 min 38 sec
Device = Mi Note -3
Camera resolution of device = 16 mp


My Code

var compressVideo = function(filePath) {
      var deferred = $q.defer();
      var videoFileName = 'TestVideo'; 
      VideoEditor.transcodeVideo(
        function (result) {
          // result is the path to the transcoded video on the device

          deferred.resolve('file://'+result);
          console.log('videoTranscodeSuccess, result: ' + 'file://' +result);
        },
        function (err) {
          deferred.reject(err);
          //console.log('videoTranscodeError, err: ' + err);
        },

        {
          fileUri: filePath,
          outputFileName: videoFileName,
          outputFileType: VideoEditorOptions.OutputFileType.MPEG4,
          optimizeForNetworkUse: VideoEditorOptions.OptimizeForNetworkUse.YES,
          saveToLibrary: true,
          videoBitrate: 1000000, // 1 megabit
          fps: 24, // optional (android only), defaults to 24
          audioChannels: 2,
          audioSampleRate: 44100,
          audioBitrate: 128000, // 128 kilobits
          progress: function(info) {
            deferred.notify(parseInt(info*100));
          }
        });
      return deferred.promise;
    };
    
I don't know is every one also observing the similar time of comression i.e 7 min 38 sec for 1 min video.
is i am doing something wrong ? Please help me and let me know if i can contibute in some way.


Thanks
